compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -305,26 +257,123 @@ bool TaggedShapeAnalyzer::verify_tag() const
+  RET_FALSE_UNLESS(all_known(_back_perm));
+  RET_FALSE_UNLESS(all_known(_out_shape));
+
+  return true;
+}
+
+/**
+ * @brief check 'Transpose-Reshape-Transpose' can be replaced by one 'Reshape'.
+ *
+ * @example
+ *  Let's explain how analyzer check Transpose-Reshape-Transpose pattern with an exact example.
+ *
+ *  Let's assume under pattern is given :
+ *
+ *      Input(1, 7, 7, 448)
+ *            |
+ *      Transpose(perm=(0, 3, 1, 2))
+ *            |
+ *      Resahape(shape=(1, 448, 49))
+ *            |
+ *      Transpose(perm=(0, 2, 1))
+ *            |
+ *      Output(1, 49, 448)
+ *
+ *  It simulates how each dimension of the tensor's shape are transformed/moved
+ *  using a member variable named '_shape'.
+ *  'tags' in _shape record the initial order of each dimension.
+ *
+ *   TIMELINE              |   _shape states :
+ *                         |
+ *   init_shape_with_tag   |    - value :   (1)   (7)   (7)   (448)
+ *                         |    - tags  :   (-)   (0)   (1)   (2)
+ *                         |
+ *   analyze_transpose     |    - value :   (1)   (448)  (7)  (7)
+ *                         |    - tags  :   (-)   (2)    (0)  (1)
+ *                         |
+ *   analyze_reshape       |    - value :   (1)   (448)   (49)
+ *                         |    - tags  :   (-)   (2)     (0, 1)
+ *                         |
+ *   anaylze_transpose     |    - value  :   (1)   (49)    (448)
+ *                         |    - tags   :   (-)   (0, 1)  (2)
+ *
+ *  After all simulation done, if tags are in same order as initial _shape,
+ *  Transpose has no effect in final shape, which they can be removed as
+ *  unnecessary Ops.
+ */
+template <loco::DataType DTYPE> bool TaggedShapeAnalyzer<DTYPE>::can_remove_transposes()
 {
-  assert(loco::must_cast<luci::CircleConst *>(f_tr->perm())->dtype() == DType);
-  assert(loco::must_cast<luci::CircleConst *>(m_rs->shape())->dtype() == DType);
-  assert(loco::must_cast<luci::CircleConst *>(b_tr->perm())->dtype() == DType);
+  auto count_not_1_dim = [](std::vector<int> &shape) -> int {
+    int count = 0;
+    for (auto i : shape)
+    {
+      if (i != 1)
+        count++;
+    }
+    return count;
+  };
 
-  const luci::CircleNode *in_tensor = loco::must_cast<luci::CircleNode *>(f_tr->a());
+  auto in_rank = count_not_1_dim(_in_shape);
+  auto out_rank = count_not_1_dim(_out_shape);
 
-  init_shape_with_tag(in_tensor);
+  if (in_rank >= out_rank)
+  {
+    reset_tag(_in_shape);
 
-  analyze_transpose<DType>(f_tr);
+    analyze_transpose(_front_perm);
 
-  if (not analyze_reshape<DType>(m_rs))
-    return false;
+    RET_FALSE_UNLESS(analyze_reshape(_mid_shape));
+
+    analyze_transpose(_back_perm);
+  }
+  else
+  {
+    reset_tag(_out_shape);
+
+    analyze_transpose(reverse_perm(_back_perm));
 
-  analyze_transpose<DType>(b_tr);
+    auto shape = extract_shape(_front_transpose);
+    RET_FALSE_UNLESS(analyze_reshape(shape));
+
+    analyze_transpose(reverse_perm(_front_perm));
```

Comment:
* if `out_rank > in_rank`, analyze it in reverse order : 
   * set_tag(output_shape)
   * analyze(back_transponse) 
   * analyze(mid_resahpe)
   * analyze(front_transpose)